File: 0078-subsets/0078-subsets.java
class Solution {
    List<List<Integer>> res;
    int[] nums;
    public List<List<Integer>> subsets(int[] nums) {
        res = new ArrayList<>();
        this.nums = nums;
        
        for (int i = 0; i <= nums.length; i++) {
            backtrack(i, 0, new LinkedList<>());
        }
        
        return res;
        
    }
    
    private void backtrack(int size, int start, LinkedList<Integer> comb) {
        if (comb.size() == size) {
            res.add(new LinkedList<>(comb));
        } else {
            for (int i = start; i < nums.length; i++) {
                comb.add(nums[i]);
                backtrack(size, i+1, comb);
                comb.removeLast();
            }
        }
    }
}

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> nums;
    vector<vector<int>> subsets(vector<int>& nums) {
        this->nums = nums;
        for (int i = 0; i <= nums.size(); i++) {
            vector<int> comb;
            backtrack(comb, i, 0);
        }
        return res;
    }
    
    void backtrack(vector<int>& comb, int size, int i) {
        if (size == comb.size()) {
            res.push_back(comb);
            return;
        } else {
            for (int j = i; j < nums.size(); j++) {
                comb.push_back(nums[j]);
                backtrack(comb, size, j+1);
                comb.pop_back();
            }
        }
    }
};